### ISSUE TITLE
Sneakers::Publisher leaves channels open while still creating new ones

### ISSUE THREAD

**Issue by veetow** [jondot/sneakers#192]
Channels never get closed.  Consider using `Bunny::Session#with_channel` for auto-closed, temporary channels.

http://www.rubydoc.info/github/ruby-amqp/bunny/master/Bunny%2FSession%3Awith_channel

**Comment by michaelklishin** [comment]
Not closed in what context? Can you please elaborate?

**Comment by veetow** [comment]
Sorry, I meant to add details yesterday, but I got busy at work.  It appears to be isolated to (maybe just the way I'm using) `Sneakers::Publisher#publish`.

In `Sneakers.configure`, I am passing in an existing `Bunny::Session`.

In my publisher class I have a class method that publishes some data on a schedule:

```
def self.publish(payload)
  Sneakers::Publisher.new(self.publish_options).publish(payload, self.publish_options)
end
```

Each time the schedule fires, a new channel is opened on the existing connection and the grow without getting closed.

**Comment by veetow** [comment]
If I do this, all is right in the world:

```
def self.publish(payload)
  p = Sneakers::Publisher.new(self.publish_options)

  p.publish(payload, self.publish_options)
  p.instance_variable_get(:@channel).close
end
```

**Comment by rud** [comment]
The Publisher class  is not really useful for anything than a quick demo of how to kind of use Bunny. Every time you create an instance you get a new connection to RabbitMQ - https://github.com/jondot/sneakers/blob/081208977eb5d606b1b7f6f3e24f83410e2c305b/lib/sneakers/publisher.rb#L10

So, a quick fix would be to reuse the instance, or create your own publisher class that can take different options for each time you publish.

**Comment by rud** [comment]
That said, the documentation for the `Publisher` class should probably highlight this, you are not the first to be burned by this code.

**Comment by veetow** [comment]
@rud, I think it can be polished up a bit to work well enough for general case stuff.  I'm working on a pull request now.  Something to the effect of this:

```
diff --git a/lib/sneakers/publisher.rb b/lib/sneakers/publisher.rb
index ec49038..719b9bd 100644
--- a/lib/sneakers/publisher.rb
+++ b/lib/sneakers/publisher.rb
@@ -12,7 +12,10 @@ module Sneakers
       to_queue = options.delete(:to_queue)
       options[:routing_key] ||= to_queue
       Sneakers.logger.info {"publishing <#{msg}> to [#{options[:routing_key]}]"}
-      @exchange.publish(msg, options)
+
+      @bunny.with_channel do |channel|
+        channel.exchange(@opts[:exchange], @opts[:exchange_options]).publish(msg, options)
+      end
     end
 
 
@@ -25,8 +28,6 @@ module Sneakers
       @bunny = @opts[:connection]
       @bunny ||= create_bunny_connection
       @bunny.start
-      @channel = @bunny.create_channel
-      @exchange = @channel.exchange(@opts[:exchange], @opts[:exchange_options])
     end
 
     def connected?
```

**Comment by veetow** [comment]
PR sent, @rud.  Let me know what you think.

**Comment by mtowers** [comment]
Here's a publisher class we created for this case:  https://github.com/RealSelf/stream-ruby/blob/master/lib/realself/stream/publisher.rb

Though it's not safe to share a single instance across workers.  If you need to do that, look in to using `ConnectionPool` as described here:  https://github.com/jondot/sneakers/issues/6#issuecomment-207507870

**Comment by michaelklishin** [comment]
Addressed by #193 AFAICT.
